src/factory.ts

```diff
@@ -21,98 +21,89 @@ import {OCaml} from './strategies/ocaml';
 import {Python} from './strategies/python';
-import {Ruby} from './strategies/ruby';
-import {RubyYoshi} from './strategies/ruby-yoshi';
+import {Ruby, RubyStrategyOptions} from './strategies/ruby';
+import {RubyYoshi, RubyYoshiStrategyOptions} from './strategies/ruby-yoshi';
 import {Rust} from './strategies/rust';
-import {Simple} from './strategies/simple';
+import {Simple, SimpleStrategyOptions} from './strategies/simple';
 import {TerraformModule} from './strategies/terraform-module';
 import {Helm} from './strategies/helm';
 import {Elixir} from './strategies/elixir';
 import {Dart} from './strategies/dart';
 import {Node} from './strategies/node';
 import {GitHub} from './github';
-import {ReleaserConfig, PluginType, RepositoryConfig} from './manifest';
-import {DefaultVersioningStrategy} from './versioning-strategies/default';
-import {VersioningStrategy} from './versioning-strategy';
+import {ReleaserConfig} from './manifest';
 import {AlwaysBumpPatch} from './versioning-strategies/always-bump-patch';
 import {ServicePackVersioningStrategy} from './versioning-strategies/service-pack';
 import {DependencyManifest} from './versioning-strategies/dependency-manifest';
-import {ManifestPlugin} from './plugin';
-import {NodeWorkspace} from './plugins/node-workspace';
-import {CargoWorkspace} from './plugins/cargo-workspace';
-import {ChangelogNotes, ChangelogSection} from './changelog-notes';
-import {GitHubChangelogNotes} from './changelog-notes/github';
-import {DefaultChangelogNotes} from './changelog-notes/default';
 import {BaseStrategyOptions} from './strategies/base';
-import {LinkedVersions} from './plugins/linked-versions';
 import {DotnetYoshi} from './strategies/dotnet-yoshi';
+import {buildVersioningStrategy} from './factories/versioning-strategy-factory';
+import {buildChangelogNotes} from './factories/changelog-notes-factory';
+
+export * from './factories/changelog-notes-factory';
+export * from './factories/plugin-factory';
+export * from './factories/versioning-strategy-factory';
 
 // Factory shared by GitHub Action and CLI for creating Release PRs
 // and GitHub Releases:
 // add any new releasers you create to this type as well as the `releasers`
 // object below.
-const allReleaseTypes = [
-  'dart',
-  'dotnet-yoshi',
-  'elixir',
-  'go',
-  'go-yoshi',
-  'helm',
-  'java-backport',
-  'java-bom',
-  'java-lts',
-  'java-yoshi',
-  'krm-blueprint',
-  'node',
-  'ocaml',
-  'php',
-  'php-yoshi',
-  'python',
-  'ruby',
-  'ruby-yoshi',
-  'rust',
-  'simple',
-  'terraform-module',
-] as const;
-export type ReleaseType = typeof allReleaseTypes[number];
-type ReleaseBuilder = (options: BaseStrategyOptions) => Strategy;
-type Releasers = Record<string, ReleaseBuilder>;
-const releasers: Releasers = {
+
+// TODO add JavaStrategyOptions
+export type AnyStrategyOptions =
+  | BaseStrategyOptions
+  | RubyStrategyOptions
+  | RubyYoshiStrategyOptions
+  | SimpleStrategyOptions;
```

Comment:
I'm wondering if I made a mistake in organizing these option types. Perhaps we should have gone with a general `StrategyOptions` which is a superset of all the possible strategy options and each strategy only picks and chooses what it uses.

We could rectify by turning `BaseStrategyOptions` into that superset of options (because it's currently part of the exported public interface) and remove the extra strategy option types. This would simplify the `ReleaseBuilder` type to always be `(options: BaseStrategyOptions) => Strategy` and we'd never have to add another type to the `AnyStrategyOptions` union type (and reduce the number of types we have to expose).

Reply:
I agree, specific options are in manifest configs anyway, so trying to split them in this specific place does not provide real loose coupling, only adds complexity.
Additionally, `SimpleStrategyOptions` is exactly same as `RubyStrategyOptions` and `RubyYoshiStrategyOptions` (all adds only `versionFile?: string` property).

So, idea is to add `versionFile` and `snapshotLabels` to `BaseStrategyOptions` and remove specific types altogether? For backward compatibility, I would just add type aliases for old names.

Reply:
To be exact, all options are mushed together in `ReleaserConfigJson` atm.